src/k2/cpo/service/Service.cpp

```diff
@@ -173,39 +173,47 @@ seastar::future<> CPOService::_assignAllTSOs() {
         });
 }
 
-seastar::future<> CPOService::_assignTSO(String &ep, size_t tsoID, int retry) {
-    auto txep = RPC().getTXEndpoint(ep);
-    K2LOG_I(log::cposvr, "assigning TSO: {}", ep);
-    if (!txep) {
-        K2LOG_W(log::cposvr, "unable to obtain endpoint for {}", ep);
-        return seastar::make_ready_future();
-    }
-    dto::AssignTSORequest request{.tsoID=tsoID+1000, .tsoErrBound=_TSOErrorBound()};
-
-    return RPC().callRPC<dto::AssignTSORequest, dto::AssignTSOResponse>(
-        dto::Verbs::GET_TSO_ASSIGN, request, *txep, _assignTimeout()
-    ).then([&ep, retry, tsoID,this] (auto &&result) {
-        auto& [status, resp] = result;
-        if (status.is2xxOK()) {
-            K2LOG_I(log::cposvr, "tso successfully assigned for endpoint: {}", ep);
-            _healthyTSOs.push_back(ep);
-            K2LOG_I(log::cposvr, "assigned TSOs: {}", _healthyTSOs);
-            return seastar::make_ready_future();
-        }
-        else if (status.is5xxRetryable() && retry > 0) {
-            K2LOG_I(log::cposvr, "cpo retrying assignment for endpoint: {} due to {}", ep, status);
-            return seastar::sleep(_assignTimeout())
-            .then([&ep, tsoID, retry, this] {
-                return _assignTSO(ep, tsoID, retry-1);
-            });
-        }
-        else {
-            K2LOG_W(log::cposvr, "tso assignment unsucessful for endpoint: {} due to {}", ep, status);
+seastar::future<> CPOService::_doAssignTSO(String &ep, size_t tsoID) {
+    return seastar::do_with(k2::dto::AssignTSORequest{.tsoID=tsoID+1000, .tsoErrBound=_TSOErrorBound()}, [&ep, tsoID, this](auto &request) {
```

Comment:
callRPC does not require the request to live beyond the calling context (it copies data to a payload), so this do_with clause is not needed. Removing it helps readability a bit